kernel/term_typing.ml

```diff
@@ -80,40 +106,21 @@ let infer_primitive env { prim_entry_type = utyp; prim_entry_content = p; } =
       in
       univs, typ
 
-    | Some (typ,Monomorphic_entry) ->
-      assert (AbstractContext.is_empty auctx); (* ensured by ComPrimitive *)
-      let u = Instance.empty in
-      let typ =
-        let typ = Typeops.infer_type env typ in
-        check_primitive_type env p u typ.utj_val;
-        typ.utj_val
-      in
-      Monomorphic, typ
-
-    | Some (typ,Polymorphic_entry uctx) ->
-      assert (not (AbstractContext.is_empty auctx)); (* ensured by ComPrimitive *)
-      (* [push_context] will check that the universes aren't repeated in
-         the instance so comparing the sizes works. No polymorphic
-         primitive uses constraints currently. *)
-      if not (AbstractContext.size auctx = UContext.size uctx
-              && Constraints.is_empty (UContext.constraints uctx))
-      then CErrors.user_err Pp.(str "Incorrect universes for primitive " ++
-                                str (op_or_type_to_string p));
-      let env = push_context ~strict:false uctx env in
-      (* Now we know that uctx matches the auctx *)
+    | Some (typ,univ_entry) ->
+      let univ_entry = adjust_primitive_univ_entry p auctx univ_entry in
+      let env, usubst, u, univs = process_universes env univ_entry in
       let typ = (Typeops.infer_type env typ).utj_val in
-      let () = check_primitive_type env p (UContext.instance uctx) typ in
-      let uctx = UContext.refine_names (AbstractContext.names auctx) uctx in
-      let u, auctx = abstract_universes uctx in
-      let typ = Vars.subst_univs_level_constr (make_instance_subst u) typ in
-      Polymorphic auctx, typ
+      let _ = check_primitive_type env p u typ in
```

Comment:
```suggestion
      let () = check_primitive_type env p u typ in
```

Reply:
Better indeed.

In passing, I found a naming mismatch and renamed `infer_declaration` to `infer_constant` to be consistent.

Reply:
@SkySkimmer: couldn't we merge `MonoTyCtx` and `PolyTyCtx`? After all, they differ only in the universes since the `usubst` of `PolyTyCtx` can be recomputed from the `auctx`?

Reply:
>the usubst of PolyTyCtx can be recomputed from the auctx?

This is not true.

We could have 
~~~ocaml
type typing_context_univs =
  | Mono
  | Poly of universe_level_subst * AbstractContext.t

type typing_context = Environ.env * unsafe_type_judgment * Id.Set.t * typing_context_univs
~~~

Reply:
> > the usubst of PolyTyCtx can be recomputed from the auctx?
> 
> This is not true.

Ok, indeed, no instance in the auctx.

> We could have
> 
> ```ocaml
> type typing_context_univs =
>   | Mono
>   | Poly of universe_level_subst * AbstractContext.t
> 
> type typing_context = Environ.env * unsafe_type_judgment * Id.Set.t * typing_context_univs
> ```

Good to me.

Reply:
I realize that by using an empty subst in the monomorphic case, we can depend on `Declarations.universe` only!